### PR TITLE
ASAN fix for test_malloc_regression

### PR DESCRIPTION
**Pull request by vlserov**
Malloc regression test checks memory leaks by comparing allocated memory before and after the test.
ASAN allocator never releases memory to OS, instead of it just marks those memory as "freed".
Generic OS functions for getting currently allocated memory aren't working correctly under ASAN.
Have to use __sanitizer_get_current_allocated_bytes function in this case.

### CODE REVIEW

File: test/common/memory_usage.h [hunk 1]
```diff
@@ -86,7 +90,9 @@ namespace utils {
     /* Returns 0 if not implemented on platform. */
     std::size_t GetMemoryUsage(MemoryStatType stat = currentUsage) {
         utils::suppress_unused_warning(stat);
-#if __TBB_WIN8UI_SUPPORT || defined(WINAPI_FAMILY)
+#if __TBB_TEST_USE_ADDRESS_SANITIZER
+        return __sanitizer_get_current_allocated_bytes();
```
Comment (alexey-katranov) [inline test/common/memory_usage.h:94]: Can you please add a comment why we need to use special functions with ASAN?

**Comment by anton-potapov** [comment]
> 
> 
> Using __sanitizer_get_current_allocated_bytes function for getting currently allocated memory under Address SANitizer

Could you please update the description what the problem is  - and how the patch is intended to fix it ?

File: test/common/memory_usage.h (hunk 1 above)
Comment (vlserov) [reply test/common/memory_usage.h:95]: Updated with comment below:
Under ASAN generic memory usage functions provide information with ASAN overhead.

File: test/common/config.h [hunk 2]
```diff
@@ -74,4 +74,10 @@ const unsigned MByte = 1024*1024;
 #define __TBB_TEST_SKIP_AFFINITY 1
 #endif
 
+#if defined(__has_feature)
+#if __has_feature(address_sanitizer)
+    #define __TBB_TEST_USE_ADDRESS_SANITIZER 1
```
Comment (anton-potapov) [inline test/common/config.h:79]: It seems `__has_feature` is the Clang-only extension and some alternative way is needed here is needed to support GCC as well

Comment (alexey-katranov) [reply test/common/config.h:79]: `__SANITIZE_ADDRESS__`
https://gcc.gnu.org/onlinedocs/gcc-4.8.0/cpp/Common-Predefined-Macros.html

**Review by alexey-katranov — DISMISSED** [review]

File: test/common/config.h [hunk 3]
```diff
@@ -74,4 +74,16 @@ const unsigned MByte = 1024*1024;
...
+	#define __TBB_HAS_FEATURE(x) __has_feature(x)
+#else
+	#define __TBB_HAS_FEATURE(x) 0
+#endif
+
+#ifdef __SANITIZE_ADDRESS__
+	#define __TBB_TEST_USE_ADDRESS_SANITIZER 1
+#else
+	#define __TBB_TEST_USE_ADDRESS_SANITIZER __TBB_HAS_FEATURE(address_sanitizer)
+#endif
```
Comment (anton-potapov) [inline test/common/config.h:87]:
```suggestion
#ifdef __SANITIZE_ADDRESS__
	#define __TBB_TEST_USE_ADDRESS_SANITIZER 1
#elif defined(__has_feature)
#if __has_feature(address_sanitizer)
      #define __TBB_TEST_USE_ADDRESS_SANITIZER 1
#endif
#endif
```

Comment (vlserov) [reply test/common/config.h:87]: Possible "undef" warning for __TBB_TEST_USE_ADDRESS_SANITIZER define in case using clang without address_sanitizer.

Comment (anton-potapov) [reply test/common/config.h:87]: could you please check it ? 
As similar check here https://github.com/oneapi-src/oneTBB/blob/master/include/oneapi/tbb/detail/_config.h#L438-L444 seems to not produce warnings

Comment (vlserov) [reply test/common/config.h:87]: It's not a real issue, since warring "undef" is switched off by default.

**Review by alexey-katranov — APPROVED** [review]

**Comment by anton-potapov** [comment]
**LGTM**

**Review by anton-potapov — APPROVED** [review]

**Pull request merged**